### ISSUE TITLE
Behaviour between slow and fast LLaMa tokenizer not equivalent

### ISSUE THREAD

**Issue by NielsRogge** [huggingface/transformers#23889]
### System Info

Transformers v4.29.2

### Who can help?

@ArthurZucker 

### Reproduction

For a new model (#23460), I'd like to get equivalent behaviour between the slow and fast LLaMa tokenizers. The code of the slow tokenizer was taken from the [original code](https://github.com/salesforce/LAVIS/blob/59273f651b9bffb193d1b12a235e909e9f826dda/lavis/models/blip2_models/blip2_vicuna_instruct.py#L82-L89), and now I'd like to translate this to the fast tokenizer as well.

However, as can be seen below, behaviour is not equivalent:

```
from transformers import LlamaTokenizer, LlamaTokenizerFast
import torch

tokenizer = LlamaTokenizer.from_pretrained("huggyllama/llama-7b", truncation_side="left")
tokenizer.add_special_tokens({"pad_token": "[PAD]"})
tokenizer.add_special_tokens({"bos_token": "</s>"})
tokenizer.add_special_tokens({"eos_token": "</s>"})
tokenizer.add_special_tokens({"unk_token": "</s>"})

fast_tokenizer = LlamaTokenizerFast.from_pretrained("huggyllama/llama-7b", truncation_side="left")
fast_tokenizer.add_special_tokens({"pad_token": "[PAD]"})
fast_tokenizer.add_special_tokens({"bos_token": "</s>"})
fast_tokenizer.add_special_tokens({"eos_token": "</s>"})
fast_tokenizer.add_special_tokens({"unk_token": "</s>"})

prompt = "What is unusual about this image?"

encoding = tokenizer(prompt, return_tensors="pt")

fast_encoding = fast_tokenizer(prompt, return_tensors="pt")

for k,v in encoding.items():
    assert torch.allclose(fast_encoding[k], v)
=> this assertion fails since the input_ids differ:

tensor([[    2,  1724,   338, 22910,  1048,   445,  1967, 29973]])
tensor([[    1,  1724,   338, 22910,  1048,   445,  1967, 29973]])
```

### Expected behavior

I'd expect that the assertion above passes.

**Comment by ArthurZucker** [comment]
Thanks for reporting, will have a look

**Comment by ArthurZucker** [comment]
Ok, the issue is in RUST, adding a token that is already in the special_tokens_map still updates something. simple reproducing script: 
```python 
In [1]: from transformers import LlamaTokenizer, LlamaTokenizerFast
   ...: 
   ...: tokenizer = LlamaTokenizer.from_pretrained("huggyllama/llama-7b", truncation_side="left")
   ...: tokenizer.add_special_tokens({"pad_token": "[PAD]"})
Out[1]: 1

In [2]: tokenizer.encode("[PAD]")
Out[2]: [1, 32000]

In [3]: tokenizer.add_special_tokens({"bos_token": "</s>"})
Out[3]: 0

In [4]: tokenizer.encode("[PAD]")
Out[4]: [2, 32000]
```

**Comment by ArthurZucker** [comment]
(this will update the processor)

**Comment by NielsRogge** [comment]
So I'm using the latest version of Transformers, have added `fast_tokenizer.bos_token = "</s>"`, but the assertion still fails for me.

**Comment by NielsRogge** [comment]
Reproduced on main branch, here's a Colab notebook: https://colab.research.google.com/drive/1KA_mliTsvjnhOCO3SApVJkgVd2HEeVQZ?usp=sharing.

**Comment by ArthurZucker** [comment]
Indeed, locally this was working, and updating the `bos_token` is supposed to change the template processor. Looking into it

**Comment by ArthurZucker** [comment]
Actually, with fast tokenizers there is no logic to properly update the template processor if it exists. The default has always been to initialize the model with the correct tokens. Meaning: 
fast_tokenizer = LlamaTokenizerFast.from_pretrained("huggyllama/llama-7b", truncation_side="left", bos_token=“</s>”)
Is what you should be using. The template processor gets updated only if you change “add_bos” and “add_eos” otherwise the logic is a bit complicated, we have to overload parent setters to bos_token as well as bos_token_id to update the template processing. Not in favor of that so leaving as is, will improve the doc for changing bos and eos in fast

**Comment by NielsRogge** [comment]
Hmm ok so there's no way to have an equivalent fast tokenizer that makes the script above pass?

The reason is that for the new InstructBLIP model (https://github.com/huggingface/transformers/pull/23460), the processor class (`InstructBlipProcessor`) would normally use the `AutoTokenizer` class to load files from the hub. And as the `AutoTokenizer` API uses the fast tokenizer by default, I'm currently not getting equivalent results as when I use the slow one.

**Comment by ArthurZucker** [comment]
No way no. I am not in favor of introducing a very hacky behaviour while the fix should be in rust in that case. 
The following should however work: 
```python 
from transformers import LlamaTokenizer, LlamaTokenizerFast
import torch

fast_tokenizer = LlamaTokenizer.from_pretrained("huggyllama/llama-7b", truncation_side="left", bos_token = "</s>", unk_token = "</s>")
fast_tokenizer.add_special_tokens({"pad_token": "[PAD]"})


tokenizer = LlamaTokenizer.from_pretrained("huggyllama/llama-7b", truncation_side="left", bos_token = "</s>", unk_token = "</s>")
tokenizer.add_special_tokens({"pad_token": "[PAD]"})
prompt = "What is unusual about this image?"

encoding = tokenizer(prompt, return_tensors="pt")

fast_encoding = fast_tokenizer(prompt, return_tensors="pt")

for k,v in encoding.items():
    assert torch.allclose(fast_encoding[k], v)
```

**Comment by ArthurZucker** [comment]
Also once you have a tokenizer ready, you can save it and it should have the correct postProcessor

**Comment by NielsRogge** [comment]
Ok thanks a lot, now works fine and I can use the fast tokenizer.

**Comment by dsdanielpark** [comment]
Hello, [ArthurZucker](https://github.com/ArthurZucker).

Thank you for the great package and maintenance. 
I wanted to inquire whether llama2's fast tokenizer is currently functioning correctly based on the above code. Or still have a problem?

**Comment by ArthurZucker** [comment]
The above message is pretty much unrelated, and should help you improve the performances when padding the input. Should work alright! 😉
